File: client/ui/updater/Download.cs
using System.Diagnostics;
using System.Net;
using System.Net.Http.Headers;

internal static class Download
{
    public static async Task FileAsync(
        string url,
        string targetPath,
        int maxParallelParts = 8,
        Action<FileDownloadProgress>? reportProgress = null,
        Action<string>? reportStatus = null,
        CancellationToken cancellationToken = default)
    {
        ArgumentOutOfRangeException.ThrowIfLessThan(maxParallelParts, 1);

        using var handler = new SocketsHttpHandler
        {
            MaxConnectionsPerServer = maxParallelParts,
            EnableMultipleHttp2Connections = true
        };
        using var client = new HttpClient(handler) { Timeout = Timeout.InfiniteTimeSpan };

        using var probeRequest = new HttpRequestMessage(HttpMethod.Get, url);
        probeRequest.Headers.Range = new(0, 0);
        using var probeResponse = await client.SendAsync(
            probeRequest,
            HttpCompletionOption.ResponseHeadersRead,
            cancellationToken);

        if (probeResponse.StatusCode != HttpStatusCode.PartialContent)
        {
            probeResponse.EnsureSuccessStatusCode();
            await DownloadResponse(probeResponse, targetPath, reportProgress, cancellationToken);
            return;
        }
        if (probeResponse.Content.Headers.ContentRange?.Length is not { } totalBytes)
        {
            await DownloadSingle(client, url, targetPath, reportProgress, cancellationToken);
            return;
        }

        try
        {
            await DownloadInParallel(
                client,
                url,
                targetPath,
                totalBytes,
                maxParallelParts,
                reportProgress,
                reportStatus,
                cancellationToken);
        }
        catch (Exception exception) when (exception is HttpRequestException or InvalidDataException or IOException)
        {
            reportStatus?.Invoke($"并行下载失败，切换为单流下载: {exception.Message}");
            await DownloadSingle(client, url, targetPath, reportProgress, cancellationToken);
        }
    }

    private static async Task DownloadSingle(
        HttpClient client,
        string url,
        string targetPath,
        Action<FileDownloadProgress>? reportProgress,
        CancellationToken cancellationToken)
    {
        using var response = await client.GetAsync(url, HttpCompletionOption.ResponseHeadersRead, cancellationToken);
        response.EnsureSuccessStatusCode();
        await DownloadResponse(response, targetPath, reportProgress, cancellationToken);
    }

    private static async Task DownloadResponse(
        HttpResponseMessage response,
        string targetPath,
        Action<FileDownloadProgress>? reportProgress,
        CancellationToken cancellationToken)
    {
        var progress = new ProgressTracker(response.Content.Headers.ContentLength, reportProgress);
        try
        {
            await using var source = await response.Content.ReadAsStreamAsync(cancellationToken);
            await using var target = File.Create(targetPath);
            var downloadedBytes = await CopyStream(source, target, progress, cancellationToken);
            if (response.Content.Headers.ContentLength is { } expectedBytes && downloadedBytes != expectedBytes)
                throw new InvalidDataException($"下载长度不正确，应为{expectedBytes}，实际为{downloadedBytes}");
        }
        finally
        {
            progress.Finish();
        }
    }

    private static async Task DownloadInParallel(
        HttpClient client,
        string url,
        string targetPath,
        long totalBytes,
        int maxParallelParts,
        Action<FileDownloadProgress>? reportProgress,
        Action<string>? reportStatus,
        CancellationToken cancellationToken)
    {
        var partCount = (int)Math.Min(maxParallelParts, totalBytes);
        var partSize = (totalBytes + partCount - 1) / partCount;
        reportStatus?.Invoke($"启动{partCount}线程高速下载");
        var partPaths = Enumerable.Range(0, partCount)
            .Select(index => $"{targetPath}.part-{index}")
            .ToArray();
        var progress = new ProgressTracker(totalBytes, reportProgress);

        try
        {
            await Task.WhenAll(partPaths.Index().Select(part =>
            {
                var start = part.Index * partSize;
                var end = Math.Min(totalBytes - 1, start + partSize - 1);
                return DownloadRange(client, url, part.Item, start, end, progress, cancellationToken);
            }));

            await using var target = File.Create(targetPath);
            foreach (var partPath in partPaths)
            {
                await using var part = File.OpenRead(partPath);
                await part.CopyToAsync(target, cancellationToken);
            }
        }
        finally
        {
            progress.Finish();
            foreach (var partPath in partPaths)
                File.Delete(partPath);
        }
    }

    private static async Task DownloadRange(
        HttpClient client,
        string url,
        string partPath,
        long start,
        long end,
        ProgressTracker progress,
        CancellationToken cancellationToken)
    {
        using var request = new HttpRequestMessage(HttpMethod.Get, url);
        request.Headers.Range = new RangeHeaderValue(start, end);
        using var response = await client.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken);
        if (response.StatusCode != HttpStatusCode.PartialContent)
            throw new HttpRequestException($"服务器未返回Range数据: {(int)response.StatusCode}");
        if (response.Content.Headers.ContentRange is not { From: var actualStart, To: var actualEnd } ||
            actualStart != start || actualEnd != end)
        {
            throw new InvalidDataException($"服务器返回了错误的Range: {response.Content.Headers.ContentRange}");
        }

        await using var source = await response.Content.ReadAsStreamAsync(cancellationToken);
        await using var target = File.Create(partPath);
        var downloadedBytes = await CopyStream(source, target, progress, cancellationToken);
        if (downloadedBytes != end - start + 1)
            throw new InvalidDataException($"分片长度不正确，应为{end - start + 1}，实际为{downloadedBytes}");
    }

    private static async Task<long> CopyStream(
        Stream source,
        Stream target,
        ProgressTracker progress,
        CancellationToken cancellationToken)
    {
        var buffer = new byte[128 * 1024];
        long downloadedBytes = 0;
        int bytesRead;
        while ((bytesRead = await source.ReadAsync(buffer, cancellationToken)) > 0)
        {
            await target.WriteAsync(buffer.AsMemory(0, bytesRead), cancellationToken);
            progress.Report(bytesRead);
            downloadedBytes += bytesRead;
        }
        return downloadedBytes;
    }

    private sealed class ProgressTracker(long? totalBytes, Action<FileDownloadProgress>? reportProgress)
    {
        private readonly Stopwatch stopwatch = Stopwatch.StartNew();
        private readonly object reportLock = new();
        private long downloadedBytes;
        private long lastReportMilliseconds = -100;

        public void Report(int bytes)
        {
            Interlocked.Add(ref downloadedBytes, bytes);
            Publish(false);
        }

        public void Finish() => Publish(true);

        private void Publish(bool completed)
        {
            lock (reportLock)
            {
                var elapsedMilliseconds = stopwatch.ElapsedMilliseconds;
                if (!completed && elapsedMilliseconds - lastReportMilliseconds < 100)
                    return;

                lastReportMilliseconds = elapsedMilliseconds;
                var downloaded = Interlocked.Read(ref downloadedBytes);
                reportProgress?.Invoke(new(
                    downloaded,
                    totalBytes,
                    downloaded / Math.Max(stopwatch.Elapsed.TotalSeconds, 0.001),
                    completed));
            }
        }
    }
}

internal readonly record struct FileDownloadProgress(
    long DownloadedBytes,
    long? TotalBytes,
    double BytesPerSecond,
    bool Completed);

File: client/updater/src/download.rs
use std::fs::File;
use std::io::{self, Read, Write};
use std::path::Path;
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::{Duration, Instant};

use crate::http::{Client, Response};
use anyhow::{Context, Result, anyhow, bail};

#[derive(Clone, Debug)]
pub struct FileDownloadProgress {
    pub downloaded_bytes: u64,
    pub total_bytes: Option<u64>,
    pub bytes_per_second: f64,
    pub completed: bool,
}

pub type ProgressCallback = Arc<dyn Fn(FileDownloadProgress) + Send + Sync>;
pub type StatusCallback = Arc<dyn Fn(String) + Send + Sync>;

pub fn file(
    url: &str,
    target_path: &Path,
    max_parallel_parts: usize,
    report_progress: Option<ProgressCallback>,
    report_status: Option<StatusCallback>,
) -> Result<()> {
    if max_parallel_parts == 0 {
        bail!("max_parallel_parts必须大于0");
    }

    let client = Client::with_timeout(Duration::ZERO);

    let response = client
        .get(url)
        .header("Range", "bytes=0-0")
        .send()
        .with_context(|| format!("请求下载探测失败: {url}"))?;

    if response.status() != 206 {
        let response = response.error_for_status().context("下载响应失败")?;
        return download_response(response, target_path, report_progress);
    }

    let total_bytes = response
        .header("Content-Range")
        .and_then(parse_content_range_length);
    drop(response);

    let Some(total_bytes) = total_bytes else {
        return download_single(&client, url, target_path, report_progress);
    };

    match download_parallel(
        &client,
        url,
        target_path,
        total_bytes,
        max_parallel_parts,
        report_progress.clone(),
        report_status.clone(),
    ) {
        Ok(()) => Ok(()),
        Err(error) => {
            if let Some(report_status) = report_status {
                report_status(format!("并行下载失败，切换为单流下载: {error}"));
            }
            download_single(&client, url, target_path, report_progress)
        }
    }
}

fn download_single(
    client: &Client,
    url: &str,
    target_path: &Path,
    report_progress: Option<ProgressCallback>,
) -> Result<()> {
    let response = client
        .get(url)
        .send()
        .with_context(|| format!("请求下载失败: {url}"))?
        .error_for_status()
        .context("下载响应失败")?;
    download_response(response, target_path, report_progress)
}

fn download_response(
    mut response: Response,
    target_path: &Path,
    report_progress: Option<ProgressCallback>,
) -> Result<()> {
    let total_bytes = response.content_length();
    let progress = ProgressTracker::new(total_bytes, report_progress);
    let result = (|| -> Result<()> {
        let mut target = File::create(target_path)
            .with_context(|| format!("创建下载文件失败: {}", target_path.display()))?;
        let downloaded_bytes = copy_stream(&mut response, &mut target, &progress)?;
        if let Some(expected_bytes) = total_bytes {
            if downloaded_bytes != expected_bytes {
                bail!("下载长度不正确，应为{expected_bytes}，实际为{downloaded_bytes}");
            }
        }
        Ok(())
    })();
    progress.finish();
    result
}

fn download_parallel(
    client: &Client,
    url: &str,
    target_path: &Path,
    total_bytes: u64,
    max_parallel_parts: usize,
    report_progress: Option<ProgressCallback>,
    report_status: Option<StatusCallback>,
) -> Result<()> {
    let part_count = max_parallel_parts.min(total_bytes as usize).max(1);
    let part_size = total_bytes.div_ceil(part_count as u64);
    if let Some(report_status) = report_status {
        report_status(format!("启动{part_count}线程高速下载"));
    }

    let part_paths: Vec<_> = (0..part_count)
        .map(|index| part_path(target_path, index))
        .collect();
    let progress = ProgressTracker::new(Some(total_bytes), report_progress);

    let result = (|| -> Result<()> {
        let mut handles = Vec::with_capacity(part_count);
        for (index, part_path) in part_paths.iter().enumerate() {
            let client = client.clone();
            let url = url.to_owned();
            let part_path = part_path.clone();
            let progress = progress.clone();
            let start = index as u64 * part_size;
            let end = (total_bytes - 1).min(start + part_size - 1);
            handles.push(thread::spawn(move || {
                download_range(&client, &url, &part_path, start, end, &progress)
            }));
        }

        let mut thread_error = None;
        for handle in handles {
            match handle.join() {
                Ok(Ok(())) => {}
                Ok(Err(error)) if thread_error.is_none() => thread_error = Some(error),
                Ok(Err(_)) => {}
                Err(_) if thread_error.is_none() => {
                    thread_error = Some(anyhow!("Range下载线程崩溃"));
                }
                Err(_) => {}
            }
        }
        if let Some(error) = thread_error {
            return Err(error);
        }

        let mut target = File::create(target_path)
            .with_context(|| format!("创建合并文件失败: {}", target_path.display()))?;
        for part_path in &part_paths {
            let mut part = File::open(part_path)
                .with_context(|| format!("打开下载分片失败: {}", part_path.display()))?;
            io::copy(&mut part, &mut target).context("合并下载分片失败")?;
        }
        Ok(())
    })();

    progress.finish();
    for part_path in part_paths {
        let _ = std::fs::remove_file(part_path);
    }
    result
}

fn download_range(
    client: &Client,
    url: &str,
    part_path: &Path,
    start: u64,
    end: u64,
    progress: &ProgressTracker,
) -> Result<()> {
    let range = format!("bytes={start}-{end}");
    let mut response = client
        .get(url)
        .header("Range", range)
        .send()
        .with_context(|| format!("Range请求失败: {start}-{end}"))?;
    if response.status() != 206 {
        bail!("服务器未返回Range数据: {}", response.status());
    }

    let content_range = response
        .header("Content-Range")
        .ok_or_else(|| anyhow!("服务器未返回Content-Range"))?;
    let (actual_start, actual_end) = parse_content_range(content_range)
        .ok_or_else(|| anyhow!("服务器返回了错误的Range: {content_range}"))?;
    if actual_start != start || actual_end != end {
        bail!("服务器返回了错误的Range: {content_range}");
    }

    let mut target = File::create(part_path)
        .with_context(|| format!("创建下载分片失败: {}", part_path.display()))?;
    let downloaded_bytes = copy_stream(&mut response, &mut target, progress)?;
    let expected_bytes = end - start + 1;
    if downloaded_bytes != expected_bytes {
        bail!("分片长度不正确，应为{expected_bytes}，实际为{downloaded_bytes}");
    }
    Ok(())
}

fn copy_stream<R: Read, W: Write>(
    source: &mut R,
    target: &mut W,
    progress: &ProgressTracker,
) -> Result<u64> {
    let mut buffer = [0u8; 128 * 1024];
    let mut downloaded_bytes = 0;
    loop {
        let bytes_read = source.read(&mut buffer).context("读取下载流失败")?;
        if bytes_read == 0 {
            break;
        }
        target
            .write_all(&buffer[..bytes_read])
            .context("写入下载文件失败")?;
        progress.report(bytes_read as u64);
        downloaded_bytes += bytes_read as u64;
    }
    Ok(downloaded_bytes)
}

fn parse_content_range_length(value: &str) -> Option<u64> {
    value.rsplit_once('/')?.1.parse().ok()
}

fn parse_content_range(value: &str) -> Option<(u64, u64)> {
    let range = value.strip_prefix("bytes ")?.split('/').next()?;
    let (start, end) = range.split_once('-')?;
    Some((start.parse().ok()?, end.parse().ok()?))
}

fn part_path(target_path: &Path, index: usize) -> std::path::PathBuf {
    std::path::PathBuf::from(format!("{}.part-{index}", target_path.display()))
}

#[derive(Clone)]
struct ProgressTracker {
    total_bytes: Option<u64>,
    report_progress: Option<ProgressCallback>,
    started_at: Instant,
    downloaded_bytes: Arc<std::sync::atomic::AtomicU64>,
    report_lock: Arc<Mutex<ReportState>>,
}

struct ReportState {
    last_report_at: Instant,
}

impl ProgressTracker {
    fn new(total_bytes: Option<u64>, report_progress: Option<ProgressCallback>) -> Self {
        Self {
            total_bytes,
            report_progress,
            started_at: Instant::now(),
            downloaded_bytes: Arc::new(std::sync::atomic::AtomicU64::new(0)),
            report_lock: Arc::new(Mutex::new(ReportState {
                last_report_at: Instant::now() - Duration::from_millis(100),
            })),
        }
    }

    fn report(&self, bytes: u64) {
        self.downloaded_bytes
            .fetch_add(bytes, std::sync::atomic::Ordering::Relaxed);
        self.publish(false);
    }

    fn finish(&self) {
        self.publish(true);
    }

    fn publish(&self, completed: bool) {
        let Some(report_progress) = &self.report_progress else {
            return;
        };
        let mut state = self.report_lock.lock().expect("下载进度锁中毒");
        let elapsed = self.started_at.elapsed();
        if !completed && state.last_report_at.elapsed() < Duration::from_millis(100) {
            return;
        }
        state.last_report_at = Instant::now();
        let downloaded = self
            .downloaded_bytes
            .load(std::sync::atomic::Ordering::Relaxed);
        report_progress(FileDownloadProgress {
            downloaded_bytes: downloaded,
            total_bytes: self.total_bytes,
            bytes_per_second: downloaded as f64 / elapsed.as_secs_f64().max(0.001),
            completed,
        });
    }
}
